torch/csrc/jit/codegen/cuda/transform_iter.cpp

```diff
@@ -440,6 +473,277 @@ int BestEffortReplay::findFirstMismatchedID(
+        // - && this is a merge
+        //   for the sake of BestEffortReplay we can forward the input mapping
+        //   to both consumer and producer to the output of the expression
+        std::vector<IterDomain*> forwarded_ids;
+        std::vector<IterDomain*> compliment_ids;
+
+        for (auto input_id : input_ids) {
+          if (!isIdOnlyInConsumer(input_id)) {
+            forwarded_ids.emplace_back(input_id);
+            forwarding_map.emplace(std::make_pair(input_id, merge_expr->out()));
+          } else {
+            compliment_ids.push_back(input_id);
+          }
+        }
+
+        // Set up compliment map
+        for (auto forwarded_id : forwarded_ids) {
+          compliment_map.emplace(std::make_pair(forwarded_id, compliment_ids));
+        }
+      }
+    }
+  }
+};
+
+} // namespace
+
+BestEffortReplay BestEffortReplay::replayCasP(
+    const TensorView* consumer,
+    const TensorView* producer,
+    int producer_compute_at_axis,
+    const RootDomainMap& root_map) {
+  if (producer_compute_at_axis < 0)
+    producer_compute_at_axis += (int)producer->nDims() + 1;
+
+  TORCH_INTERNAL_ASSERT(
+      producer_compute_at_axis >= 0 &&
+          (unsigned int)producer_compute_at_axis <= producer->nDims(),
+      "Invalid axis provided to BestEffortReplay::replayCasP.");
+
+  // producer ids we need to match in consumer
+  std::vector<IterDomain*> producer_CA_ids(
+      producer->domain()->domain().begin(),
+      producer->domain()->domain().begin() + producer_compute_at_axis);
+  producer_CA_ids = TensorDomain::noReductions(producer_CA_ids);
+
+  // If producer has an rfactor root, that's what will match the consumer
+  std::vector<IterDomain*> producer_root = producer->getMaybeRFactorDomain();
+
+  // Figure out all inputs required to generate the compute_at dimensions. We
+  // need all deps because inputs on producer may be in getRootDomain, but we
+  // may need in rFactorDomain
+  auto all_CA_id_deps = DependencyCheck::getAllValsBetween(
+      {producer_root.begin(), producer_root.end()},
+      {producer_CA_ids.begin(), producer_CA_ids.end()});
+
+  // Figure out minimal set of root IDs needed to produce producer_CA_ids:
+  std::unordered_set<IterDomain*> producer_CA_root_ids;
+  for (IterDomain* id : producer_root) {
+    if (std::find(all_CA_id_deps.begin(), all_CA_id_deps.end(), id) !=
+        all_CA_id_deps.end()) {
+      producer_CA_root_ids.emplace(id);
+    }
+  }
+
+  const auto p2c_root_map = root_map.mapProducerToConsumer(
+      producer->domain(), consumer->domain(), producer_CA_root_ids);
+
+  ConsumerForwardingInfo consumer_forwarding_info(producer, consumer);
+
+  auto consumer_replay = BestEffortReplay(
+      consumer->domain()->domain(),
+      producer_CA_ids,
+      p2c_root_map,
+      consumer_forwarding_info.forwarding_map);
+
+  // Need to adjust leaf map based on forwarding before returning.
+
+  // ID's could go through more than one forward iteration in the map before it
```

Comment:
This logic was quite wrong before. @naoyam could you review it again, just collecting all forwarded ids.

Reply:
LGTM.